### DISCUSSION TITLE
Wrong Calculation - Feed-in Values & Grid Consumption Values don't make sense (with a three phase - H3) inverter) !

### DISCUSSION THREAD

**Discussion by mafischerhei** [nathanmarlor/foxess_modbus#456] [General]
I think there must be something wrong with the calculation of values you get from a 3-phase setup:
At the moment I am consuming power from the net, but the values show:

Grid CT R -   -98W
Grid CT S-    -83W
Grid CT T-    +54W
-> -127 (seems ok)

Grid Consumption R -    +98W
Grid Consumption S -    +83W
Grid Consumption T -      +0W
-> 181 (Too much! exactly: 54 too much, should be 127)

Feed-In R -   +0W
Feed-In S -   +0W
Feed-In T -   +54W
-> 54W (should be ZERO! or minus 127)

I think the "Grid CT" combined is plausible, the rest is not.

I think the calculation of the entities of "Grid Consumption" and "Feed-In" is calculated wrong in the code

it's NOT possible with 3-phases to just take the "positive" values (Grid Consumption / Feed-In) and ignore negative values. It always has to be a calculation (negative AND positive values).

I think there are only two ways to get correct values for this:

Either:
Allow negative values for feed-IN and Grid Consumption and keep R,S,T for those 6 entities:

Grid CT R -   -98W
Grid CT S-    -83W
Grid CT T-    +54W
-> -127W

Grid Consumption R -    +98W
Grid Consumption S -    +83W
Grid Consumption T -     -54W (!)
-> 127W

Feed-In R -   -98W (!)
Feed-In S -   -83W (!)
Feed-In T -   +54W
-> -127W

OR:
simply using a combined entity in the code:
Grid Consumption (combined) - sum up all "Grid Consumption" no matter if they are positive or negative and then, make this value positive if negative or vice-versa (or make it Zero when it makes sense
Same for feed-in

What do you think, guys?

**Comment by canton7** [comment]
Which calculation exactly do you think is incorrect? We don't provide a "Grid Consumption", "Feed-in" or "Grid CT" entity for the H3.

**Reply by canton7** [reply]
Note that Grid Consumption and Feed-In are derived directly from Grid CT: if the value of a phase is >0 it's counted as Feed-in, and if it's <0 it's counted as Grid Consumption. The inverter only supplies us with Grid CT.

So by definition Feed-in and Grid Consumption values cannot be negative.

**Reply by mafischerhei** [reply]
hmm. I have:
Gerät Informationen
H3-8.0-E - AUX
von FoxESS


yes, i DO have : Grid CT R, Grid CT S, Grid CT T, Grid Consumption R , Grid Consumption S, Grid Consumption T, Feed-In R, Feed-In S, Feed-In T
from this integration
?
And I am happy they're there....

Or do you mean you don't provide COMBINED entities for "Grid Consumption", "Feed-in" or "Grid CT"? That's clear to me.

**Reply by canton7** [reply]
I'm unclear what exactly you think the "wrong calculation" is. 

The Grid CT values are read directly from the inverter. There's no calculation there.

The Grid Consumption and Feed-in are just exact same values from Grid CT, but Feed-in contains the positive values and Grid Consumption contains the negative values.

**Reply by mafischerhei** [reply]
@canton7 I agree: Feed-In and Grid Consumption can not be negative.
Maybe it's a question of one's point of view.
My idea was to make "combined" sensors (I dont' care for a single phase, ... what for? I want to know how much I export or import and not what a single phase imports or exports. what use case should that be for?)
one for Grid CT, one for Grid Consumption and one for  Feed-in
for Grid CT that works fine, for the others not.
Ok, maybe then the solution is to simply make my own combined sensor for Grid Consumption and Feed-In out of the Grid CT R,S,T

But honestly. It still feels wrong.
If you say, "The Grid CT values are read directly from the inverter.", same should be for Grid Consumption and Feed-In:
letting them go negative or positive, exactly like they are read from the inverter would make sense, if not, then its a kind of calculation (preventing them go negative) , which is completely useless in a 3-phase-surrounding.

( It perfectly  makes sense in a one-phase setup if consumption and Feed-In never go negative, but not in a 3-phase one)
Please. Have a look again and maybe think it over:

Grid CT R - -98W
Grid CT S- -83W
Grid CT T- +54W
-> -127 (seems ok)

Grid Consumption R - +98W
Grid Consumption S - +83W
Grid Consumption T - +0W
-> 181 (Too much! exactly: 54 too much, should be 127) - No, I DO NOT consume 181, I consume 127. The values are correct for the single phases, but the 0 gives a wrong impression.

Feed-In R - +0W
Feed-In S - +0W
Feed-In T - +54W
-> 54W (should be ZERO! or minus 127) - correct for Phase T, but nonsense if you look at them all three together, because.... still im not exporting....

**Reply by mafischerhei** [reply]
In other words: if "Grid Consumption" and Feed-In" wouldn't be prevented from going negative, and just be passed through like "Grid CT", it would look like this:

Grid CT R - -98W
Grid CT S- -83W
Grid CT T- +54W
-> -127W

Grid Consumption R - +98W
Grid Consumption S - +83W
Grid Consumption T - -54W (!)
-> 127W

Feed-In R - -98W (!)
Feed-In S - -83W (!)
Feed-In T - +54W
-> -127W

it would feel  more "logic" to me if I look at the circuit as whole.

And also the question is: what's the sense of of "Feed-in" and " Grid Consumption" in a 3-phase setup if you can't calculate with them correctly? Is there any?

**Reply by mafischerhei** [reply]
Maybe one point more to convince you when you say that  Grid Consumption and Feed-In R can not get negative. (which is absolutely true if you see it "as whole"!)
You could argue similar for the Grid CT: It can be POSITIVE of NEGATIVE, but not both at the same time. But we DO have positive and negative values at the same time. Because we're in a 3-phase setup. We have to make an addition of 3 phases to get useful values. Always. Same should be for all sensors. 

Or even better: Some combined sensors. If Grid Consumption + Feed-In would be already combined sensors in your code. Then, of course, it would make sense again preventing them from going negative

**Reply by canton7** [reply]
> And also the question is: what's the sense of of "Feed-in" and " Grid Consumption" in a 3-phase setup if you can't calculate with them correctly? Is there any?

Yeah, I think those were introduced to match the H1, before we understood exactly how the H3 behaved regarding the different phases.

I guess the most obvious solution is to just remove Grid Consumption and Feed-in, but 1) people might be using them, so that's a breaking change, and 2) you can do that yourself by just disabling the entities.

**Reply by TonyM1958** [reply]
Not exactly - Grid Consumption and Feedin are still needed for 3 phase as this is the way net metering works with the energy supplier. You are not charged individually for import and export on each phase - you are charged for your net (in most countries - they are some that are different as noted above). So, You sum the RPower, SPower and TPower and then split this on sign into Grid Consumption and Feedin exactly as you do for single phase.

grid_daily and feedin_daily are then (normally) used to populate the values on the Energy Dashboard. If you don't generate these, what will people use here?

**Reply by TonyM1958** [reply]
There is an exception here where I don't sum RST power to get net RST power - when calculating inveter efficiency from inverter power in and power. Here you need to include -ve RST power values in power in and +ve RST values in power out separately or the inverter efficiency that is reported is wrong.

**Reply by mafischerhei** [reply]
@TonyM1958 For me (3-phase setup) for the energy-dashboard, those sensors (already there) work perfectly :
Feed-in Total (foxess)
Grid Consumption Total (foxess)
Battery Discharge Total (foxess)
Battery Charge Total (foxess)
PV1 Power Total (foxess)
PV2 Power Total (foxess)

**Reply by mafischerhei** [reply]
@canton7 

> I guess the most obvious solution is to just remove Grid Consumption and Feed-in, but 1) people might be using them, so that's a breaking change, and 2) you can do that yourself by just disabling the entities.

Ok, now I understand your consideration.
Ok, I KNOW how to calculate everything I need out of the "Grid CT R S T" values.
It's just .. I am someone who likes things neatly in place.
I understand that its a breaking change.
So, maybe one solution would be giving 3-phase-setups ADDITIONALLY already combined sensors  for "Feed-In R S T" and "Grid Consumption R S T", wouldn't clutter up home assistant of people already using those mention sensors.

But I understand If you have more important things to improve.

It's just a suggestion for improvement. 

Anyway, thanks for having your plugin ,-))

**Reply by TonyM1958** [reply]
The total sensors are directly from the inverter and are worked out by the firmware. They are not calculated by the integration. Neither are the Today sensors. Just because you don't use the _daily sensors does not mean others don't. This is not a custom integration just for you...

**Reply by mafischerhei** [reply]
@TonyM1958 
I think you misunderstood. It's good to have _daily sensors of course. It's clear that the total and today sensors are not calculated. I didn't question that.
I just wanted to answer your question what people should use for the dashboard, and
I can tell that total values work perfectly, because I use them for over a year now. (from various integrations)

**Reply by canton7** [reply]
@mafischerhei Can you please be a bit less shouty? We're all giving up our free time here, and having unpleasant interactions makes us disinclined to spend our time on this issue. Thanks!

**Comment by TonyM1958** [comment]
There are lots of side issues here you may not be aware with your specific configuration and this has already been discussed a lot.

To be clear, I am not using Nathan's integration currently (so the sensor name do not match), but the following code is offered as an example of something that has been run on H1, KH and H3 with main and solar inverters and the feedback is that it produces the correct load power when CT2 is correctly connected:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/ab88b491-f596-4d6f-bd94-7a3d0276281e)

An untested use case is when multiple inverters are in Master / Slave setups. Here, it is not currently clear if the master reports the net power generated / consumed by the master + slaves or whether you need a second meter to report the power flowing into / out of the slave inverters in order to get the correct load power.

All of the power values are signed  - no values are clipped, so the grid, ct2 and inverter powers can be both +ve and -ve.

On 3 phase, you see times where the inverter is importing on one phase and exporting on other phases due to the way it manages the current across the phases. For most people, 3 phase uses net metering (power values are summed across the phases) but there are issues where each phase is metered separately - which I think is the case in Hungary or Bulgaria (sorry, don't recall exactly).

'ct2_handling' is just a switch to give an option to ignore CT2 for debugging - usually CT2 is enabled in this template.

The final 0.3 just attempts to avoid -ve spikes in load power if there is a sampling error in the power values due to fast transients - by definition, load power cannot be negative.

Here's an example of what this looks like for single phase:

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/05422334-c325-4cf5-94ef-7791297bc00b)

Hope this helps?

**Reply by canton7** [reply]
> and the feedback is that it produces the correct load power when CT2 is correctly connected

What does "corectly connected" mean here? What is it connected to?

**Reply by mafischerhei** [reply]
> On 3 phase, you see times where the inverter is importing on one phase and exporting on other phases due to the way it manages the current across the phases. For most people, 3 phase uses net metering (power values are summed across the phases

Isn't that exactly what I am saying? ( talking still about 3-phase setups):

How this integration gives out properties right now is like:
Grid CT R - -98W
Grid CT S- -83W
Grid CT T- +54W
-> -127 (seems ok) - correct

Grid Consumption R - +98W - correct for the single phase
Grid Consumption S - +83W - correct for the single phase
Grid Consumption T - +0W - correct for the single phase (not allowed to give "-" values), but:
all three phases summed up  together -> 181 WRONG if you sum up together, I don't import THAT much.

Feed-In R - +0W - correct for the single phase  (not allowed to give "-" values)
Feed-In S - +0W - correct for the single phase  (not allowed to give "-" values)
Feed-In T - +54W - correct for the single phase, but:
all three phases summed up  together -> 54W WRONG  if you sum up together, because there's no export at all.

I think it should be on 3 phases calculated like this

Grid Consumption= Grid Consumption R + Grid Consumption S + Grid Consumption T (Letting allow negative or positive Results for the single phases, and then, AFTER calculation allow only positive values for the RESULT.
Would be in my case: (98)+(83)+(-54) = 127 , which would be correct. 

That's why I am saying, if "Grid Consumption R S T" and "Feed-In R S T" , the single phases are not allowed to go negative , they're pretty useless for a 3-phase setup, because they have to be summed up before restricting the RESULT of the [calculation] from going negative.

**Reply by mafischerhei** [reply]
@TonyM1958 Thanks for the code. Brings me closer to calculating the correct Houseload
Where I live I would say I have a pretty typical configuration H3+Chint
Which side issues do I have? Not clear what you mean.

Is your example code about the Inverter Power or the Houseload? Asking because of the unique name in this example.


But I have a little trouble translating the example into the entities of this integration here:
sensor.eps_rpower
sensor.eps_spower
sensor.eps_tpower
are clear

sensor.rpower, sensor.spower, sensor.tpower would translate to sensor.foxess_inv_power_r/s/t
 I assume?

sensor.grid_ct : Would that be the combination of sensor.foxess_grid_ct_r/s/t ? and if yes:
What would be the  sensor.meter_r/s/tpower in this integration here?
Isn't it the same as sensor.grid_ct_r/s/t ?

we don't have single entities for ct1 / ct2 in this integration here. Do we?

**Reply by TonyM1958** [reply]
The code works on H1, KH and H3 inverters. The different sensors for these models are described here: https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Modbus-Map.xls

The code that I use to configure the sensors for the different inverter models and the templates that you are copying is here: https://github.com/TonyM1958/HA-FoxESS-Modbus

**Comment by TonyM1958** [comment]
> > and the feedback is that it produces the correct load power when CT2 is correctly connected
> 
> What does "corectly connected" mean here? What is it connected to?

Correctly connected means it is measuring the output power that is being injected into the local grid supply from all secondary inverters and is attached the right way around.

**Reply by canton7** [reply]
Ah OK, you're assuming the presence of extra inverters

**Reply by TonyM1958** [reply]
Yes - or conversely, I'm not assuming extra inverters are not present!

**Reply by TonyM1958** [reply]
BTW - you only get CT2/Meter for H3 - there are no separate phases for this. So, the load power is only correct when you present it as a virtual single phase inverter.
